### DISCUSSION TITLE
Netvips DzSave Linux X64 Output How to control Zoomlevels

### DISCUSSION THREAD

**Discussion by hkvaka** [libvips/libvips#3162] [General]
Hi,

I am using NetVips library with C# and .NET 6 deploying the application on a Linux X64  Azure appservice.
the DzSave errored initially as it could not find libvips.so.42 file in the deployed location after installing the nuget package 
"NetVips.Native.linux-x64" Version="8.12.2" the application started to work and produce output files.

"NetVips" Version="2.2.0"

However the same application with the same input file when ran on windows produces higher zoom levels.
but the linux version not producing even 10th of windows version.

Could some one suggest what parameters can be tweaked to get better zoom levels in linux

Input Image used
https://openslide.cs.cmu.edu/download/openslide-testdata/Mirax/CMU-1-Saved-1_2.zip

output directories from windows are 0-17
![image](https://user-images.githubusercontent.com/118275347/201931313-59ce8292-1023-4b34-858a-733cd641308a.png)

output directories on linux are 0-11
![image](https://user-images.githubusercontent.com/118275347/201931646-a59fb18a-12f8-4468-aa65-b567c8e78d2d.png)

Windows X64 uses additional runtime libraries from https://github.com/libvips/build-win64-mxe

Windows Shows Total number of pels to process = **6732422740**

Linux Total number of pels to process = **826880**

![image](https://user-images.githubusercontent.com/118275347/202227948-6f9f5e82-bc08-43bf-b906-4ebdbb4a2ac8.png)

Regards
Hari

**Comment by hkvaka** [comment]
Any comments on the above problem, am new to linux and not sure if any further packages need to be deployed on the Azure Function App Service for Linux 
![image](https://user-images.githubusercontent.com/118275347/202394833-5eef04b3-8153-411a-939b-c4a9b04c77c0.png)

**Comment by kleisauke** [comment]
> "NetVips.Native.linux-x64" Version="8.12.2"

Hello, the `NetVips.Native.*` NuGet packages does not support OpenSlide, you need to make sure that libvips is compiled with OpenSlide support to avoid that MRXS images are being loaded with the TIFF loader. See: https://github.com/kleisauke/net-vips/issues/18#issuecomment-427906129.

**Reply by hkvaka** [reply]
HI kleisauke,

Many thanks for pointing me to the blog, looks like that is focused on windows and my windows setup currently works in my local dev setup, 

Apologies I am new to Linux it will be really helpful If you could expand on how openslide support can be integrated with linux. 
or point me to any installation documentation. I do not have physical access to the linux machine, Its a azure function app.

![image](https://user-images.githubusercontent.com/118275347/202408307-e3d941b3-cad3-4268-8501-93d894f61be1.png)

**Comment by hkvaka** [comment]
I am sort of blocked with Linux environment getting working, it will be very helpful to know what steps are involved in getting the DzSave working with open slide support on Linux target machine with .NET 6 executable

**Comment by kleisauke** [comment]
Never having used Azure functions myself, but I'm aware that installing Linux packages via the package manager (i.e. apt in this case) can be done on Azure functions by either:
- Using an initialization script to install the required libraries on the application (function) startup (which may exceed the startup command execution time limit);
- Deploying your function using a custom Linux container, but [it would require Premium or Dedicated plan](https://learn.microsoft.com/en-us/azure/azure-functions/functions-create-function-linux-custom-image).

Without testing, I think the first one can be done like this:
```bash
apt-get update &&
DEBIAN_FRONTEND=noninteractive apt-get install -y libvips-dev --no-install-recommends &&
dotnet <YourApplicationNameHere>.dll
```` 

and the latter by:
```Dockerfile
FROM mcr.microsoft.com/dotnet/sdk:6.0 AS installer-env

COPY . /src/dotnet-function-app
RUN cd /src/dotnet-function-app && \
    mkdir -p /home/site/wwwroot && \
    dotnet publish *.csproj --output /home/site/wwwroot

# To enable SSH and remote debugging on app service change the base image to the one below
# FROM mcr.microsoft.com/azure-functions/dotnet-isolated:4.0-dotnet-isolated6.0-appservice
FROM mcr.microsoft.com/azure-functions/dotnet-isolated:4.0-dotnet-isolated6.0

ENV DEBIAN_FRONTEND=noninteractive \
    AzureWebJobsScriptRoot=/home/site/wwwroot \
    AzureFunctionsJobHost__Logging__Console__IsEnabled=true

RUN apt-get update \
    && apt-get install -y libvips-dev --no-install-recommends

COPY --from=installer-env ["/home/site/wwwroot", "/home/site/wwwroot"]
```

However, I cannot recommend this for production use because Azure functions run on Debian 11 (bullseye), so it will end up installing libvips 8.10.5.

**Comment by hkvaka** [comment]
kleisauke,

Thankyou for your suggestion to use debian package store to install the libvips package
It was really helpful and now I could progress further using the linux hosting, however some of the mrxs files that we intend to use are around 5-10 GB some cases even bigger.

I have noticed a 3.65 GB file taking around 12 hours processing the DzSave, is this reasonable also could you suggest any parameters that we need to pass to the DzSave which can control how many output files it generates I have noticed that the output directories having more than a million files generated.

can control the zoom levels to 50%, 75% that will reduce the final output files and the time to process without compromising on quality
